core/src/Kokkos_SYCL.hpp

```diff
@@ -83,7 +83,7 @@ class SYCL {
   SYCL();
   explicit SYCL(const sycl::queue&);
 
-  uint32_t impl_instance_id() const noexcept { return 0; }
+  uint32_t impl_instance_id() const noexcept { return 1; }
```

Comment:
So, the way this is happening is that there are two classes of ExecutionSpace: those that are differentiable, and those that aren't. In the case of SYCL, there's only one instance, so impl_instace_id is set to 1. This is changing because "0" now means something special, an operation on "0" is happening on _all instances_. A sync on Cuda + 0 is a deviceSynchronize.

Also this isn't the first time impl_instance_id has been used, that gets used by Profiling already to determine which dev_id to pass to the tool

Reply:
Yes, in `SYCL` no queue is special.

Reply:
I think I'm still confused what "In the case of SYCL, there's only one instance" means to you. How is `SYCL` different from `CUDA` or `HIP` in this regard?

Reply:
Actually, looking at this more, I think SYCL is just wired differently than the others. I was looking here: https://github.com/kokkos/kokkos/blob/master/core/src/SYCL/Kokkos_SYCL_Instance.hpp#L58-L67 , and saw no constructor from a sycl::queue, which made me think all SYCL instances submit to the same queue. But that's not true, is it? It's just that in our case, the SYCL queue is stored in the SYCL instance, not the SYCLInternal. Is there a way to get a unique ID for a SYCL queue? That's probably what we need.

Reply:
Yes, an execution space can be initialized with a `sycl::queue`, see https://github.com/kokkos/kokkos/blob/7b5ffcc3224c5d4fa8ef46dd390c59f2d55752e8/core/src/Kokkos_SYCL.hpp#L84 and it is used for initializing an `SYCLInternal` object which stores a copy of it, see https://github.com/kokkos/kokkos/blob/7b5ffcc3224c5d4fa8ef46dd390c59f2d55752e8/core/src/SYCL/Kokkos_SYCL_Instance.hpp#L89. Hence, I would think that we can use the same approach as for CUDA and HIP.

Reply:
Okay. So something to the effect of "hash my SYCLInternal object ptr"? Sorry for the slow movement here, just trying to make sure we get the right thing

Reply:
Yes, AFAICT, we do the same for `CUDA` and `HIP`, i.e. the execution space classes store a pointer to `CudaInternal` resp. `HIPInternal` which contains a `cudaStream_t` resp. `hipStream_t` member.
So I would expect
```C++
uint32_t SYCL::impl_instance_id() const noexcept {
  return m_space_instance->impl_get_instance_id();
}

uint32_t SYCLInternal::impl_get_instance_id() const {
  return Kokkos::Tools::Experimental::Impl::idForInstance<Kokkos::Experimental::SYCL>(
      reinterpret_cast<uintptr_t>(this));
}
```